src/linear_algebra/determinant.lean

```diff
@@ -52,6 +54,50 @@ variables {M' : Type*} [add_comm_group M'] [module R M']
 variables {ι : Type*} [decidable_eq ι] [fintype ι]
 variables (e : basis ι R M)
 
+section conjugate
+
+variables {A : Type*} [integral_domain A]
+variables {m n : Type*} [fintype m] [fintype n]
+
+
+/-- If `R^m` and `R^n` are linearly equivalent, then `m` and `n` are also equivalent. -/
+def equiv_of_pi_lequiv_pi {R : Type*} [integral_domain R]
+  (e : (m → R) ≃ₗ[R] (n → R)) : m ≃ n :=
+basis.index_equiv (basis.of_equiv_fun e.symm) (pi.basis_fun _ _)
+
+/-- If `M` and `M'` are each other's inverse matrices, they are square matrices up to
+equivalence of types. -/
+def matrix.index_equiv_of_inv [decidable_eq m] [decidable_eq n]
+  {M : matrix m n A} {M' : matrix n m A}
+  (hMM' : M ⬝ M' = 1) (hM'M : M' ⬝ M = 1) :
+  m ≃ n :=
+equiv_of_pi_lequiv_pi (matrix.to_lin'_of_inv hMM' hM'M)
+
+/-- If `M'` is a two-sided inverse for `M` (indexed differently), `det (M ⬝ N ⬝ M') = det N`. -/
+lemma matrix.det_conjugate
+  [decidable_eq m] [decidable_eq n]
+  {M : matrix m n A} {M' : matrix n m A} {N : matrix n n A}
+  (hMM' : M ⬝ M' = 1) (hM'M : M' ⬝ M = 1) :
+  det (M ⬝ N ⬝ M') = det N :=
+begin
+  letI := classical.dec_eq A,
+  -- Although `m` and `n` are different a priori, we will show they have the same cardinality.
+  -- This turns the problem into one for square matrices, which is easy.
+  let e : m ≃ n := matrix.index_equiv_of_inv hMM' hM'M,
+  have : det (matrix.reindex_linear_equiv e (equiv.refl _) M ⬝ N ⬝
+              matrix.reindex_linear_equiv (equiv.refl _) e M') =
+         det N,
+  { rw [det_mul, det_mul, mul_comm, ← mul_assoc, ← det_mul, ← matrix.reindex_linear_equiv_mul,
+        matrix.reindex_linear_equiv_refl_refl, hM'M, linear_equiv.refl_apply, det_one, one_mul] },
+  convert this,
+  rw [← matrix.det_reindex_linear_equiv_self e (M ⬝ N ⬝ M'),
+      matrix.reindex_linear_equiv_mul e (equiv.refl n) e,
+      matrix.reindex_linear_equiv_mul e (equiv.refl n) (equiv.refl n),
+      matrix.reindex_linear_equiv_refl_refl, linear_equiv.refl_apply]
+end
```

Comment:
I was able to show this using the lemmas in #7685 with:
```suggestion
begin
  -- Although `m` and `n` are different a priori, we will show they have the same cardinality.
  -- This turns the problem into one for square matrices (`matrix.det_units_conj`), which is easy.
  let e : m ≃ n := matrix.index_equiv_of_inv hMM' hM'M,
  let U : units (matrix n n A) := ⟨
    M.minor e.symm (equiv.refl _),
    M'.minor (equiv.refl _) e.symm,
    by rw [mul_eq_mul, ←minor_mul_equiv, hMM', minor_one_equiv],
    by rw [mul_eq_mul, ←minor_mul_equiv, hM'M, minor_one_equiv]⟩,
  rw [
    ←matrix.det_units_conj U N,
    ←det_minor_equiv_self e.symm,
    minor_mul_equiv _ _ _ (equiv.refl _) _,
    minor_mul_equiv _ _ _ (equiv.refl _) _],
  refl,
end
```
Do you think this is better or worse than your proof?

Reply:
I would like to see one proved in terms of the other, although it doesn't matter too much which one is first. The way I see it, either we push all the complexity to the "hard" `det_conjugate` case and let `det_units_conj` be a direct corollary, or we can do the "easy" `det_units_consj` case first, then show in `det_conjugate` that the hard case reduces to the easy case. Neither is obviously better than the other to me.

Reply:
In the interest of getting things in the PR queue moving again, I'm merging your PR and rebasing mine on top of it.